sonar-core/src/main/java/org/sonar/core/issue/tracking/Tracker.java

```diff
@@ -0,0 +1,291 @@
+        tracking.associateManualIssueToLine(base, null);
+      } else {
+        String baseHash = base.getLineHash();
+        if (Strings.isNullOrEmpty(baseHash)) {
+          baseHash = baseInput.getLineHashSequence().getHashForLine(base.getLine());
+        }
+        if (!Strings.isNullOrEmpty(baseHash)) {
+          int[] rawLines = rawInput.getLineHashSequence().getLinesForHash(baseHash);
+          if (rawLines.length == 1) {
+            tracking.associateManualIssueToLine(base, rawLines[0]);
+          } else if (rawLines.length == 0 && rawInput.getLineHashSequence().hasLine(base.getLine())) {
+            // still valid (???). We didn't manage to correctly detect code move, so the
+            // issue is kept at the same location, even if code changes
+            tracking.associateManualIssueToLine(base, base.getLine());
+          }
+          // TODO if hash found multiple times, , pick the closest line
+        }
+      }
+    }
+  }
+
+  private enum IsManual implements Predicate<Trackable> {
+    INSTANCE;
+    @Override
+    public boolean apply(Trackable input) {
+      return input.getRuleKey().isManual();
+    }
+  }
+
+  private interface SearchKey {
+  }
+
+  private interface SearchKeyFactory {
+    SearchKey create(Trackable trackable);
+  }
+
+  private static class LineAndLineHashKey implements SearchKey {
+    private final RuleKey ruleKey;
+    private final String lineHash;
+    private final Integer line;
+
+    LineAndLineHashKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.line = trackable.getLine();
+      this.lineHash = StringUtils.defaultString(trackable.getLineHash(), "");
+    }
+
+    @Override
+    public boolean equals(@Nonnull Object o) {
+      if (this == o) {
+        return true;
+      }
+      LineAndLineHashKey that = (LineAndLineHashKey) o;
+      // start with most discriminant field
+      return Objects.equals(line, that.line)
+        && lineHash.equals(that.lineHash)
+        && ruleKey.equals(that.ruleKey);
+    }
+
+    @Override
+    public int hashCode() {
+      int result = ruleKey.hashCode();
+      result = 31 * result + lineHash.hashCode();
+      result = 31 * result + (line != null ? line.hashCode() : 0);
+      return result;
+    }
+  }
+
+  private enum LineAndLineHashKeyFactory implements SearchKeyFactory {
+    INSTANCE;
+    @Override
+    public SearchKey create(Trackable t) {
+      return new LineAndLineHashKey(t);
+    }
+  }
+
+  private static class LineHashAndMessageKey implements SearchKey {
+    private final RuleKey ruleKey;
+    private final String message;
+    private final String lineHash;
+
+    LineHashAndMessageKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.message = trackable.getMessage();
+      this.lineHash = StringUtils.defaultString(trackable.getLineHash(), "");
+    }
+
+    @Override
+    public boolean equals(@Nonnull Object o) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Add a type test to this method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2097)